### ISSUE TITLE
"docker commit" on Azure (TP5) fails with: hcsshim::ImportLayer failed in Win32

### ISSUE THREAD

**Issue by mathiasconradt** [MicrosoftDocs/Virtualization-Documentation#294]
On Azure with TP5, I try to create a new image out of an existing container.

It fails with the error:

```
Error response from daemon: hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\b4b3a30008
f1e3b1fbb0f7dda8c3ec027e71564ab3a45a8e1622765d8732c5f5 flavour=1 folder=C:\Windows\TEMP\hcs960930749
```

What I did via cmd:

```
PS C:\Windows\system32> docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
company/mycompany       latest              e081b9d046b7        9 days ago          12.11 GB
windowsservercore   10.0.14300.1000     dbfee88ee9fd        9 weeks ago         9.344 GB
windowsservercore   latest              dbfee88ee9fd        9 weeks ago         9.344 GB

PS C:\Windows\system32> docker ps -a
CONTAINER ID        IMAGE               COMMAND                   CREATED             STATUS                         PORTS                NAMES
6a4f1ec207bf        company/mycompany       "cmd /S /C 'C:\\company-"   9 days ago          Exited (1067) 25 minutes ago                        mycompany

PS C:\Windows\system32> docker commit mycompany company/mycompany2
Error response from daemon: hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\b4b3a30008
f1e3b1fbb0f7dda8c3ec027e71564ab3a45a8e1622765d8732c5f5 flavour=1 folder=C:\Windows\TEMP\hcs960930749
```

Looks as if other users also have this issue:
- https://social.msdn.microsoft.com/Forums/en-US/88b5f6c2-e454-4181-96c2-910651d3eff0/docker-commit-problem-hcsshimimportlayer-failed-in-win32-the-system-cannot-find-the-path?forum=windowscontainers
- https://social.msdn.microsoft.com/Forums/en-US/f759a5ab-4f18-4b16-84fa-8246a4292a40/docker-commit-issue?forum=windowscontainers

docker version:

```
PS C:\> docker version
Client:
 Version:      1.12.0-dev
 API version:  1.24
 Go version:   go1.5.3
 Git commit:   2b97201
 Built:        Tue Apr 26 23:51:36 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.0-dev
 API version:  1.24
 Go version:   go1.5.3
 Git commit:   2b97201
 Built:        Tue Apr 26 23:51:36 2016
 OS/Arch:      windows/amd64
```

**Comment by friism** [comment]
@mathiasconradt I'm getting this too, did you find a solution?

**Comment by mathiasconradt** [comment]
@friism No, never got this to work on TP5. I found other issues in TP5, so I stopped evaluating Docker on TP5 since June, waiting for the final release (afaik planned to be out next week) that hopefully fixes the remaining issues that I'm having.

**Comment by StefanScherer** [comment]
I've tested this on Azure with Docker 1.12.1 and it worked. I've updated Docker with [this script](https://github.com/StefanScherer/docker-windows-box/blob/bea002f59ec57522c5ea51708998348a86ddb519/scripts/update-docker-1.12.1.ps1)

```
PS C:\Users\vagrant> docker run -it --name test microsoft/windowsservercore cmd
Microsoft Windows [Version 10.0.14300]
(c) 2016 Microsoft Corporation. All rights reserved.

C:\>echo hello >hello.txt

C:\>exit
PS C:\Users\vagrant> docker images
REPOSITORY                    TAG                 IMAGE ID            CREATED             SIZE
microsoft/windowsservercore   10.0.14300.1030     02cb7f65d61b        3 months ago        7.764 GB
microsoft/windowsservercore   latest              02cb7f65d61b        3 months ago        7.764 GB
PS C:\Users\vagrant> docker ps -a
CONTAINER ID        IMAGE                         COMMAND             CREATED             STATUS                     PORTS               NAMES
8f0894747318        microsoft/windowsservercore   "cmd"               2 minutes ago       Exited (0) 8 seconds ago                       test
PS C:\Users\vagrant> docker commit test stefanscherer/test
sha256:c0de927a221010069d1b750235116c86cd7a8b9e30baeee23cbdc9834bd935c9
PS C:\Users\vagrant> docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      windows/amd64
PS C:\Users\vagrant>
```

**Comment by dtma007** [comment]
I am still seeing this same issue on Windows Server 2016 GA (10.0.14393).  

Client:
 Version:      17.03.1-ee-3
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   3fcee33
 Built:        Thu Mar 30 19:31:22 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.03.1-ee-3
 API version:  1.27 (minimum version 1.24)
 Go version:   go1.7.5
 Git commit:   3fcee33
 Built:        Thu Mar 30 19:31:22 2017
 OS/Arch:      windows/amd64
 Experimental: false

when trying to "docker commit":
"hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\8965301e9e9735141b479ed3cc873107ef16c4bfbb3dc0e54d0b4827d034722a flavour=1 folder=C:\windows\TEMP\hcs510325079"

In my case I'm starting off from a "microsoft/mssql-server-windows-express" as my base Docker container and installing some additional software in the container.

I have tried various things such as:

   * stopping and removing any other containers on the host machine
   * rebooting
   * retrying several times

But nothing has worked

**Comment by enderb-ms** [comment]
Is this still an issue?

**Comment by mback2k** [comment]
Yes, I am also experiencing this issue while installing msys2. Use the following Dockerfile to create a base image:
```
# escape=`

FROM microsoft/windowsservercore:1709_KB4043961

SHELL ["powershell", "-command"]

RUN Invoke-WebRequest "http://www.7-zip.org/a/7z1604-x64.exe" -OutFile 7z1604-x64.exe;
RUN Start-Process -FilePath "C:\7z1604-x64.exe" -ArgumentList /S -NoNewWindow -PassThru -Wait;
RUN Remove-Item 7z1604-x64.exe;

RUN Invoke-WebRequest "http://repo.msys2.org/distrib/msys2-x86_64-latest.tar.xz" -OutFile msys2-x86_64-latest.tar.xz;
RUN Start-Process -FilePath "C:\Program` Files\7-Zip\7z.exe" -ArgumentList e, msys2-x86_64-latest.tar.xz -NoNewWindow -PassThru -Wait;
RUN Start-Process -FilePath "C:\Program` Files\7-Zip\7z.exe" -ArgumentList x, msys2-x86_64-latest.tar, `-oC:\ -NoNewWindow -PassThru -Wait;
RUN Remove-Item msys2-x86_64-latest.tar.xz; Remove-Item msys2-x86_64-latest.tar;

RUN Write-Host 'Updating MSYSTEM and MSYSCON ...'; `
	[Environment]::SetEnvironmentVariable('MSYSTEM', 'MSYS2', [EnvironmentVariableTarget]::Machine); `
	[Environment]::SetEnvironmentVariable('MSYSCON', 'defterm', [EnvironmentVariableTarget]::Machine);

# For some reason bash.exe has to be called first twice since we are not building interactive.
RUN C:\msys64\usr\bin\bash.exe -l -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe -l -c 'exit 0';
```

After that run the container using:
` docker run -i msys2-test` (do not use --rm or -t)

Inside the container perform the following steps:
`C:\msys64\usr\bin\bash.exe -l -c 'pacman -Syu --noconfirm';`
`C:\msys64\usr\bin\bash.exe -l -c 'pacman -Su --noconfirm';`

Then exit and try to commit the container. You will then get the following:
`PS E:\msys2-test> docker commit 316a96035f9e msys2-test
Error response from daemon: re-exec error: exit status 1: output: time="2017-11-16T21:52:11+01:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\d23bf18e82f5340169acd0523e561a5334be64b7bfbd24bed7b6149922635a9a flavour=1 folder=C:\\Windows\\TEMP\\hcs626213039"
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\d23bf18e82f5340169acd0523e561a5334be64b7bfbd24bed7b6149922635a9a flavour=1 folder=C:\Windows\TEMP\hcs626213039`

You can also use this complete Dockerfile to perform the manual steps using the builder which will run into the same issue.
```
# escape=`

FROM microsoft/windowsservercore:1709_KB4043961

SHELL ["powershell", "-command"]

RUN Invoke-WebRequest "http://www.7-zip.org/a/7z1604-x64.exe" -OutFile 7z1604-x64.exe;
RUN Start-Process -FilePath "C:\7z1604-x64.exe" -ArgumentList /S -NoNewWindow -PassThru -Wait;
RUN Remove-Item 7z1604-x64.exe;

RUN Invoke-WebRequest "http://repo.msys2.org/distrib/msys2-x86_64-latest.tar.xz" -OutFile msys2-x86_64-latest.tar.xz;
RUN Start-Process -FilePath "C:\Program` Files\7-Zip\7z.exe" -ArgumentList e, msys2-x86_64-latest.tar.xz -NoNewWindow -PassThru -Wait;
RUN Start-Process -FilePath "C:\Program` Files\7-Zip\7z.exe" -ArgumentList x, msys2-x86_64-latest.tar, `-oC:\ -NoNewWindow -PassThru -Wait;
RUN Remove-Item msys2-x86_64-latest.tar.xz; Remove-Item msys2-x86_64-latest.tar;

RUN Write-Host 'Updating MSYSTEM and MSYSCON ...'; `
	[Environment]::SetEnvironmentVariable('MSYSTEM', 'MSYS2', [EnvironmentVariableTarget]::Machine); `
	[Environment]::SetEnvironmentVariable('MSYSCON', 'defterm', [EnvironmentVariableTarget]::Machine);

# For some reason bash.exe has to be called first twice since we are not building interactive.
RUN C:\msys64\usr\bin\bash.exe -l -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe -l -c 'exit 0';

RUN C:\msys64\usr\bin\bash.exe -l -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe -l -c 'pacman -Syu --noconfirm';

RUN C:\msys64\usr\bin\bash.exe -l -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe -l -c 'pacman -Su  --noconfirm';
```

Also see the following Github issue: https://github.com/moby/moby/issues/32838

I need to install msys2 in order to containerize my CI setups. Please do not hesitate to ask for more information. Thanks!

**Comment by mback2k** [comment]
Attached you will find the debug logs from Windows Server 1709 using Docker version 17.10.0-ee-preview-3: [docker-debug-logs-294.zip](https://github.com/MicrosoftDocs/Virtualization-Documentation/files/1481672/docker-debug-logs-294.zip)

The error line is: `Cleaning up layer 1828dd2de551692ca2ba02662d62324d63a74212dc46011fd64915ad832720f4: re-exec error: exit status 1: output: time=""2017-11-17T09:56:24+01:00"" level=error msg=""hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\1828dd2de551692ca2ba02662d62324d63a74212dc46011fd64915ad832720f4 flavour=1 folder=C:\\Windows\\TEMP\\hcs247125035""
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\1828dd2de551692ca2ba02662d62324d63a74212dc46011fd64915ad832720f4 flavour=1 folder=C:\Windows\TEMP\hcs247125035`

**Comment by StefanScherer** [comment]
I can reproduce the error with the `Dockerfile` given above on 1709
Kernel Version: 10.0 16299 (16299.15.amd64fre.rs3_release.170928-1534)

**Comment by ljt91** [comment]
Was hitting this issue myself, found this article about increasing the container base size which resolved it for me.
https://stackoverflow.com/questions/46539165/docker-windows-build-fails-with-error-the-system-cannot-find-the-path-specifie

**Comment by enderb-ms** [comment]
Thank you guys for responding. I'm currently setting up a repo so I can share it with the team. I will reply back with any updates, or questions we may have.

**Comment by compulim** [comment]
I am hitting the very same problem. My VSTS was building Docker images but one day it fail on its own.

Every time, Docker fail roughly after about 10-11 minutes. My image is a big file copy operation (~4GB of files going in), so I am guessing maybe my Dockerfile takes too long time to complete.

**Comment by enderb-ms** [comment]
We are currently tracking this issue and have opened a bug for it. I will post back here as we find out more about the cause. In the mean time, if anyone can try running the docker build or docker commit commands using Hyper-V isolation, I'd be interested to see if that gets you different results. You can do this for docker build by adding the flag "--isolation=hyperv". For the docker commit command, run that same flag when starting the container.

**Comment by compulim** [comment]
@enderb-ms 

We tried the `--isolation=hyperv` in our VSTS build, but still get the very same error. Somehow we think it might be related to timeout because the error always occur around 10-11 minutes.

```
2017-12-06T19:53:31.9789229Z ##[section]Starting: Build hosted image
2017-12-06T19:53:31.9814064Z ==============================================================================
2017-12-06T19:53:31.9814817Z Task         : Docker
2017-12-06T19:53:31.9815644Z Description  : Build, tag, push, or run Docker images, or run a Docker command. Task can be used with Docker or Azure Container registry.
2017-12-06T19:53:31.9816698Z Version      : 0.3.2
2017-12-06T19:53:31.9817293Z Author       : Microsoft Corporation
2017-12-06T19:53:31.9817992Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=848006)
2017-12-06T19:53:31.9818797Z ==============================================================================
2017-12-06T19:53:32.5456165Z [command]"C:\Program Files\Docker\docker.exe" build -f d:\a\r1\a\Build\Dockerfile-hosted --build-arg "--isolation hyperv" -t xxx.azurecr.io/xxx:27656 -t xxx.azurecr.io/xxx d:\a\r1\a\Build
2017-12-06T19:58:59.4236538Z Sending build context to Docker daemon  2.821GB
2017-12-06T19:58:59.4246143Z 
2017-12-06T19:59:17.9238244Z Step 1/2 : FROM xxx.azurecr.io/xxx
2017-12-06T19:59:17.9248995Z  ---> 466b41d61c71
2017-12-06T19:59:17.9253181Z Step 2/2 : ADD xxx C:\\Source\\xxx
2017-12-06T20:09:04.5592569Z re-exec error: exit status 1: output: time="2017-12-06T20:08:12Z" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\edd3ba9af34d9aaa560e3abbb57a097d15417be8e3dd249b87eba9174c4066db flavour=1 folder=C:\\Windows\\TEMP\\hcs428293703"
2017-12-06T20:09:04.5598330Z hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\edd3ba9af34d9aaa560e3abbb57a097d15417be8e3dd249b87eba9174c4066db flavour=1 folder=C:\Windows\TEMP\hcs428293703
2017-12-06T20:09:04.5997404Z ##[error]C:\Program Files\Docker\docker.exe failed with return code: 1
2017-12-06T20:09:04.6064479Z ##[section]Finishing: Build hosted image
```

**Comment by mback2k** [comment]
For me `--isolation=hyperv` also does not change anything and I can re-produce the same error on Windows 10 (which only has Hyper-V based support for containers).

**Comment by jhowardmsft** [comment]
Issue from @mback2k https://github.com/MicrosoftDocs/Virtualization-Documentation/issues/294#issuecomment-345062593 fixed by https://github.com/Microsoft/hcsshim/pull/154. I suspect there may be multiple issues.

**Comment by jhowardmsft** [comment]
I'm going to close this issue due to its age, and that one fix has been submitted to hcsshim (still needs a re-vendor PR into moby/moby which I'll do once the hcsshim fix is merged). 

For others hitting this error, please open a new separate issue (rather than adding a me-too) in the moby/moby repo and provide:
- Windows build information
- Docker info/version output
- Dockerfile which reproduces the issue
- Associated files needed by the dockerfile

**Comment by compulim** [comment]
Link to the moby/moby PR to track the status.

https://github.com/moby/moby/pull/35924
